Experiment not starting: [CRITICAL] Unexpected error: 'NoneType' object has no attribute 'set'.

### Description
Half of the time I am trying to run my exp (`a9sg`), I need to re-create the exp multiple times to get rid of an error that seems to be randomly caused, or I don't fully understand. See info below, thanks in advance!
* Autosubmit version: v4.1.14
* Machine/VM/environment name (if applicable): hub5
* Experiment ID (if applicable): a9sg
* Experiment tasks & log path if applicable): /gpfs/scratch/bsc32/bsc954883/a9sg/LOG_a9sg

### Reproducible Example
You can clone autosubmit exp `a9sg` and try to create and run it to reproduce (very likely) the bug.

### Expected Behaviour
It will 50% of the time after doing `create exp` or running `scripts/prepare_autosubmit.sh` create this error:
```
[ERROR] Arguments provided: Namespace(version=False, logfile='DEBUG', logconsole='WARNING', command='run', expid='a9sg', notransitive=False, update_version=False, start_time=None, start_after=None, run_only_members=None, profile=False)
[ERROR] This is the experiment: <a9sg> which had an issue with the command: <run> and it is currently using the Autosubmit Version: 4.1.14.
 [CRITICAL] Unexpected error: 'NoneType' object has no attribute 'set'.
 Please report it to Autosubmit Developers through Git
More info at https://autosubmit.readthedocs.io/en/master/troubleshooting/error-codes.html
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,7 @@ release.
 - Fixed bug where an RO-Crate file would include itself in the archive, as well as other zip files.
   Now Autosubmit uses the pattern $expid-crate-$date-$time-$millisecond.zip, and ignores any ZIP files
   in the tmp/ASLOGS that start with $expid-crate and end with .zip #2692
+- Fixed 'NoneType' object has no attribute 'set' that would have set a 'NoneType' instead of a 'EventType' #2611 #2583
 
 **Enhancements:**
 

--- a/autosubmit/config/configcommon.py
+++ b/autosubmit/config/configcommon.py
@@ -62,17 +62,17 @@ class AutosubmitConfig(object):
         self.last_experiment_data = {}
         self.data_loops = set()
 
-        self.current_loaded_files = dict()
+        self.current_loaded_files: dict = {}
         self.conf_folder_yaml = Path(BasicConfig.LOCAL_ROOT_DIR, expid, "conf")
         if not Path(BasicConfig.LOCAL_ROOT_DIR, expid, "conf").exists():
             raise IOError(f"Experiment {expid}/conf does not exist")
         self.wrong_config = defaultdict(list)
         self.warn_config = defaultdict(list)
-        self.dynamic_variables = dict()
-        self.special_dynamic_variables = dict()  # variables that will be substituted after all files is loaded
-        self.starter_conf = dict()
+        self.dynamic_variables: dict = {}
+        self.special_dynamic_variables: dict = {}  # variables that will be substituted after all files is loaded
+        self.starter_conf: dict = {}
         self.misc_files = []
-        self.misc_data = list()
+        self.misc_data: dict = {}
         self.default_parameters = {'d': '%d%', 'd_': '%d_%', 'Y': '%Y%', 'Y_': '%Y_%',
                                    'M': '%M%', 'M_': '%M_%', 'm': '%m%', 'm_': '%m_%'}
 
@@ -117,7 +117,7 @@ class AutosubmitConfig(object):
             wrapper = {}
         return wrapper.get('EXPORT', self.experiment_data.get("WRAPPERS", {}).get("EXPORT", ""))
 
-    def get_project_submodules_depth(self) -> list:
+    def get_project_submodules_depth(self) -> list[int]:
         """Returns the max depth of submodule at the moment of cloning.
 
         Default is -1 (no limit).
@@ -131,8 +131,7 @@ class AutosubmitConfig(object):
             depth = [int(x) for x in unparsed_depth.split(",")]
         else:
             try:
-                depth = int(unparsed_depth)
-                depth = [depth]
+                depth = [int(unparsed_depth)]
             except TypeError:
                 Log.warning("PROJECT_SUBMODULES_DEPTH is not an integer neither a int. Using default value -1")
                 depth = []
@@ -906,7 +905,7 @@ class AutosubmitConfig(object):
             self,
             parameters: dict[str, Any] = None,
             max_deep: int = 25,
-            dict_keys_type: str = None,
+            dict_keys_type: str = '',
             in_the_end: bool = False,
     ) -> dict[str, Any]:
         """
@@ -931,7 +930,7 @@ class AutosubmitConfig(object):
         if parameters is None:
             parameters = self.deep_parameters_export(self.experiment_data, self.default_parameters)
 
-        if dict_keys_type is None:
+        if dict_keys_type == '':
             dict_keys_type = self.check_dict_keys_type(parameters)
 
         while len(dynamic_variables) > 0 and max_deep > 0:
@@ -2497,7 +2496,7 @@ class AutosubmitConfig(object):
 
     # based on https://github.com/cbirajdar/properties-to-yaml-converter/blob/master/properties_to_yaml.py
     @staticmethod
-    def ini_to_yaml(root_dir: Path, ini_file: str) -> None:
+    def ini_to_yaml(root_dir: Path, ini_file: Path) -> None:
         # Based on http://stackoverflow.com/a/3233356
         def update_dict(original_dict: dict, updated_dict: collections.abc.Mapping) -> dict:
             for k, v in updated_dict.items():
@@ -2532,7 +2531,7 @@ class AutosubmitConfig(object):
         config_dict = ConfigObj(input_file, stringify=True, list_values=False, interpolation=False, unrepr=False)
 
         # Store the result in yaml_dict
-        yaml_dict = {}
+        yaml_dict: dict = {}
 
         for key, value in config_dict.items():
             config_keys = key.split(".")

--- a/autosubmit/job/job_list.py
+++ b/autosubmit/job/job_list.py
@@ -1859,17 +1859,15 @@ class JobList(object):
             return [job for job in completed_jobs if job.packed is False]
         return completed_jobs
 
-    def get_completed_failed_without_logs(self, platform: Any = None) -> List[Any]:
-        """Returns a list of completed or failed jobs without updated logs.
+    def get_completed_failed_without_logs(self) -> list[Any]:
+        """
+        Returns a list of completed or failed jobs without updated logs.
 
-        :param platform: Job platform, defaults to None.
-        :type platform: Platform
         :return: List of completed and failed jobs without updated logs.
         :rtype: List[Job]
         """
 
         completed_failed_jobs = [job for job in self._job_list if
-                                 (platform is None or job.platform.name == platform.name) and
                                  (job.status == Status.COMPLETED or job.status == Status.FAILED) and
                                  job.updated_log is False]
 

--- a/autosubmit/platforms/platform.py
+++ b/autosubmit/platforms/platform.py
@@ -17,14 +17,13 @@
 
 import atexit
 import multiprocessing
-from multiprocessing.context import SpawnContext
 import os
 import queue  # only for the exception
 import time
 import traceback
 from contextlib import suppress
-from multiprocessing.synchronize import Event
 from multiprocessing.queues import Queue
+from multiprocessing.synchronize import Event
 # noinspection PyProtectedMember
 from os import _exit  # type: ignore
 from pathlib import Path
@@ -42,6 +41,7 @@ if TYPE_CHECKING:
     from autosubmit.job.job import Job
     from autosubmit.job.job_list import JobList
     from autosubmit.job.job_package_persistence import JobPackagePersistence
+    from multiprocessing.process import BaseProcess
 
 
 def _init_logs_log_process(as_conf: 'AutosubmitConfig', platform_name: str) -> None:
@@ -113,7 +113,7 @@ class CopyQueue(Queue):
         self.timeout = timeout
         super().__init__(maxsize, ctx=ctx)
 
-    def put(self, job: Any, block: bool = True, timeout: float = None) -> None:
+    def put(self, job: Any, block: bool = True, timeout: Optional[float] = None) -> None:
         """Puts a job into the queue if it is not a duplicate.
 
         :param job: The job to be added to the queue.
@@ -131,7 +131,7 @@ class Platform:
     Class to manage the connections to the different platforms.
     """
     # This is a list of the keep_alive events, used to send the signal outside the main loop of Autosubmit
-    worker_events = list()
+    worker_events: list[Event] = []
     # Shared lock between the main process and a retrieval log process
     lock = multiprocessing.Lock()
 
@@ -148,9 +148,10 @@ class Platform:
         :param auth_password: Optional password for two-factor authentication.
         :type auth_password: str or list, optional
         """
+        self.ctx = self.get_mp_context()
         self.connected = False
-        self.expid = expid  # type: str
-        self._name = name  # type: str
+        self.expid: str = expid
+        self._name: str = name
         self.config = config
         self.tmp_path = os.path.join(
             self.config.get("LOCAL_ROOT_DIR", ""), self.expid, self.config.get("LOCAL_TMP_DIR", ""))
@@ -209,12 +210,12 @@ class Platform:
             self.pw = None
         self.max_waiting_jobs = 20
         self.recovery_queue: Optional[Queue] = None
-        self.work_event = None
-        self.cleanup_event = None
-        self.log_retrieval_process_active = False
-        self.log_recovery_process = None
+        self.work_event: Optional[Event] = None
+        self.cleanup_event: Optional[Event] = None
+        self.log_retrieval_process_active: bool = False
+        self.log_recovery_process: Optional['BaseProcess'] = None
         self.keep_alive_timeout = 60 * 5  # Useful in case of kill -9
-        self.processed_wrapper_logs = set()
+        self.processed_wrapper_logs: set[str] = set()
         self.compress_remote_logs = False
         self.remote_logs_compress_type = "gzip"
         self.compression_level = 9
@@ -231,7 +232,7 @@ class Platform:
             self.compress_remote_logs = platform_config.get("COMPRESS_REMOTE_LOGS", False)
             self.remote_logs_compress_type = platform_config.get("REMOTE_LOGS_COMPRESS_TYPE", "gzip")
             self.compression_level = platform_config.get("COMPRESSION_LEVEL", 9)
-            
+
         self.log_queue_size = log_queue_size
         self.remote_log_dir = None
 
@@ -857,18 +858,18 @@ class Platform:
         This method sets the cleanup event to signal the log recovery process to finish,
         waits for the process to join with a timeout, and then resets all related variables.
         """
-        if self.cleanup_event:
+        if self.cleanup_event is not None:
             self.cleanup_event.set()  # Indicates to old child ( if reachable ) to finish.
-        if self.log_recovery_process:
+        if self.log_recovery_process is not None:
             # Waits for old child ( if reachable ) to finish. Timeout in case of it being blocked.
             self.log_recovery_process.join(timeout=60)
         # Resets everything related to the log recovery process.
-        self.recovery_queue = None
+        self.recovery_queue = self.ctx.Queue()
         self.log_retrieval_process_active = False
         self.remove_workers(self.work_event)
-        self.work_event = None
-        self.cleanup_event = None
-        self.log_recovery_process = None
+        self.work_event = self.ctx.Event()
+        self.cleanup_event = self.ctx.Event()
+        self.log_recovery_process = self.ctx.Process()
         self.processed_wrapper_logs = set()
 
     def load_process_info(self, platform):
@@ -899,23 +900,23 @@ class Platform:
             if not isinstance(self.config[key], dict) or key in ["PLATFORMS", "EXPERIMENT", "DEFAULT", "CONFIG"]:
                 platform.config[key] = self.config[key]
 
-    def prepare_process(self, ctx) -> 'Platform':
+    def prepare_process(self) -> 'Platform':
         new_platform = self.create_a_new_copy()
-        self.work_event = ctx.Event()
-        self.cleanup_event = ctx.Event()
+        self.work_event = self.ctx.Event()
+        self.cleanup_event = self.ctx.Event()
         Platform.update_workers(self.work_event)
         self.load_process_info(new_platform)
         if self.recovery_queue:
             del self.recovery_queue
         # Retrieval log process variables
-        self.recovery_queue = CopyQueue(ctx=ctx)
+        self.recovery_queue = CopyQueue(ctx=self.ctx)
         # Cleanup will be automatically prompt on control + c or a normal exit
         atexit.register(self.send_cleanup_signal)
         atexit.register(self.close_connection)
         return new_platform
 
-    def create_new_process(self, ctx: SpawnContext, new_platform: 'Platform', as_conf) -> None:
-        self.log_recovery_process = ctx.Process(
+    def create_new_process(self, new_platform: 'Platform', as_conf) -> None:
+        self.log_recovery_process = self.ctx.Process(
             target=recover_platform_job_logs_wrapper,
             args=(new_platform, self.recovery_queue, self.work_event, self.cleanup_event, as_conf),
             name=f"{self.name}_log_recovery")
@@ -942,16 +943,16 @@ class Platform:
                                                                                      "false")).lower() == "false"):
             if as_conf and as_conf.misc_data.get("AS_COMMAND", "").lower() == "run":
                 self.log_retrieval_process_active = True
-                ctx = self.get_mp_context()
-                new_platform = self.prepare_process(ctx)
-                self.create_new_process(ctx, new_platform, as_conf)
+                self.ctx = self.get_mp_context()
+                new_platform = self.prepare_process()
+                self.create_new_process(new_platform, as_conf)
                 self.join_new_process()
 
     def send_cleanup_signal(self) -> None:
         """Sends a cleanup signal to the log recovery process if it is alive.
         This function is executed by the atexit module
         """
-        if self.log_recovery_process and self.log_recovery_process.is_alive():
+        if self.log_recovery_process.is_alive():
             self.work_event.clear()
             self.cleanup_event.set()
             self.log_recovery_process.join(timeout=60)
@@ -1003,7 +1004,8 @@ class Platform:
                 break
         return process_log
 
-    def recover_job_log(self, identifier: str, jobs_pending_to_process: set[Any], as_conf: 'AutosubmitConfig') -> None:
+    def recover_job_log(self, identifier: str, jobs_pending_to_process: set[Any],
+                        as_conf: 'AutosubmitConfig') -> set[Any]:
         """Recovers log files for jobs from the recovery queue and retries failed jobs.
 
         :param identifier: Identifier for logging purposes.
@@ -1061,7 +1063,7 @@ class Platform:
         identifier = f"{self.name.lower()}(log_recovery):"
         try:
             Log.info(f"{identifier} Starting...")
-            jobs_pending_to_process = set()
+            jobs_pending_to_process: set = set()
             self.connected = False
             self.restore_connection(as_conf, log_recovery_process=True)
             Log.result(f"{identifier} successfully connected.")

--- a/test/integration/conftest.py
+++ b/test/integration/conftest.py
@@ -657,3 +657,24 @@ def as_db(request: 'FixtureRequest', autosubmit: Autosubmit, tmp_path: 'LocalPat
     autosubmit_exp('____')
 
     return backend
+
+
+def wait_child(timeout, retry=3):
+    """A parametrized fixture that will retry function X amount of times waiting for a child process to be executed.
+
+    In case it still fails after X retries an exception is thrown."""
+    def the_real_decorator(function):
+        def wrapper(*args, **kwargs):
+            retries = 0
+            while retries < retry:
+                try:
+                    value = function(*args, **kwargs)
+                    if value is None:
+                        return
+                except Exception:
+                    time.sleep(timeout)
+                    retries += 1
+
+        return wrapper
+
+    return the_real_decorator

--- a/test/integration/git/test_autosubmit_git.py
+++ b/test/integration/git/test_autosubmit_git.py
@@ -20,7 +20,7 @@
 from contextlib import nullcontext as does_not_raise
 from getpass import getuser
 from pathlib import Path
-from typing import Callable, ContextManager, Generator, Tuple, Union, cast
+from typing import Callable, ContextManager, Generator, Union
 
 import pytest
 from test.integration.conftest import DockerContainer
@@ -177,7 +177,7 @@ def test_git_submodules_dirty(
     expected to fail, raising an error when the experiment is operational.
     """
 
-    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str # type: ignore
 
     git_repo = git_repos_path / 'git_repository'
     git_submodule = git_repos_path / 'git_submodule'
@@ -251,7 +251,7 @@ def test_git_operational_experiment_toggle_flag(
     mocked_log = mocker.patch('autosubmit.autosubmit.Log')
     expid = 'o001'
 
-    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str # type: ignore
 
     test_name = request.node.name
     git_repo = git_repos_path / f'git_repository_{test_name}'
@@ -288,9 +288,9 @@ def test_git_operational_experiment_toggle_flag(
         assert exit_code == expected
 
         assert mocked_log.warning.called
-        assert mocked_log.warning.call_args[0][0] == 'Git operational check disabled by user'
+        assert mocked_log.warning.mock_calls[1][1][0] == 'Git operational check disabled by user'
     else:
-        with pytest.raises(cast(expected, Exception)):
+        with pytest.raises(Exception):
             autosubmit.run_command(args=args)
 
 
@@ -333,7 +333,7 @@ def test_clean_git_not_committed(
         git_server: Generator[tuple[DockerContainer, Path, str], None, None]
 ):
     """Test that cleaning Git fails when the project directory has new files not committed yet."""
-    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str  # type: ignore
 
     git_repo = git_repos_path / test_clean_git_not_committed.__name__
     create_git_repository(git_repo, bare=True)
@@ -362,7 +362,7 @@ def test_clean_git_not_pushed(
         git_server: Generator[tuple[DockerContainer, Path, str], None, None]
 ):
     """Test that cleaning Git fails when the project directory has staged changed not pushed."""
-    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str  # type: ignore
 
     git_repo = git_repos_path / test_clean_git_not_pushed.__name__
     create_git_repository(git_repo, bare=True)
@@ -389,10 +389,10 @@ def test_clean_git_not_pushed(
 def test_clean_git(
         tmp_path,
         autosubmit_exp,
-        git_server: Generator[Tuple[DockerContainer, Path, str], None, None]
+        git_server: Generator[tuple[DockerContainer, Path, str], None, None]
 ):
     """Test that cleaning Git fails when the project commit cannot be recorded."""
-    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str  # type: ignore
 
     git_repo = git_repos_path / test_clean_git_not_pushed.__name__
     create_git_repository(git_repo, bare=True)

--- a/test/integration/scripts/test_stats.py
+++ b/test/integration/scripts/test_stats.py
@@ -18,11 +18,13 @@
 """Test for the ``autosubmit stats`` command."""
 
 from pathlib import Path
+from time import time, sleep
+
+from psutil import Process
 
 from autosubmit.scripts.autosubmit import main
 
 _EXPID = 't000'
-
 
 def test_autosubmit_commands_help(autosubmit_exp, mocker):
     """Test that the monitor is called for stats with a simple job list.
@@ -44,7 +46,20 @@ def test_autosubmit_commands_help(autosubmit_exp, mocker):
         exp.as_conf,
         exp.expid,
         'run')
+
+    processes_before_run = Process().children(recursive=True)
     assert 0 == exp.autosubmit.run_experiment(_EXPID)
+    processes_after_run = Process().children(recursive=True)
+
+    before = time()
+    wait_n_seconds = 30
+    while time() - before < wait_n_seconds:
+        if len(processes_after_run) <= len(processes_before_run):
+            break
+        # Children processes of ``autosubmit run`` may still be running, and
+        # we need those to finish in this test so that we have all the remote
+        # data copied, and ``autosubmit stats`` can run and use those files.
+        sleep(1)
 
     mocker.patch('sys.argv', ['autosubmit', 'stats', '-o', 'png', '--section_summary',
                               '--jobs_summary', '--hide', _EXPID])

--- a/test/integration/test_log_recovery.py
+++ b/test/integration/test_log_recovery.py
@@ -24,10 +24,10 @@ from pathlib import Path
 import pytest
 
 from autosubmit.autosubmit import Autosubmit
+from autosubmit.config.configcommon import AutosubmitConfig
 from autosubmit.job.job import Job
 from autosubmit.job.job_common import Status
 from autosubmit.platforms.platform import CopyQueue
-from autosubmit.config.configcommon import AutosubmitConfig
 
 
 def _get_script_files_path() -> Path:
@@ -145,9 +145,8 @@ def test_wait_until_timeout(prepare_test, local, as_conf, mocker, cleanup_event,
     mocker.patch('autosubmit.platforms.platform.Platform.get_mp_context', return_value=mp.get_context('fork'))
     local.keep_alive_timeout = 2
     max_items = 1
-    ctx = local.get_mp_context()
-    local.prepare_process(ctx)
-    local.recovery_queue = CopyQueue(ctx=ctx)
+    local.prepare_process()
+    local.recovery_queue = CopyQueue(ctx=local.ctx)
     local.cleanup_event.set() if cleanup_event else local.cleanup_event.clear()
     local.work_event.set() if work_event else local.work_event.clear()
     if recovery_queue_full:
@@ -173,9 +172,8 @@ def test_wait_for_work(prepare_test, local, as_conf, mocker, cleanup_event, work
     mocker.patch('autosubmit.platforms.platform.Platform.get_mp_context', return_value=mp.get_context('fork'))
     local.keep_alive_timeout = 2
     max_items = 1
-    ctx = local.get_mp_context()
-    local.prepare_process(ctx)
-    local.recovery_queue = CopyQueue(ctx=ctx)
+    local.prepare_process()
+    local.recovery_queue = CopyQueue(ctx=local.ctx)
     local.cleanup_event.set() if cleanup_event else local.cleanup_event.clear()
     local.work_event.set() if work_event else local.work_event.clear()
     if recovery_queue_full:
@@ -201,9 +199,8 @@ def test_wait_mandatory_time(prepare_test, local, as_conf, mocker, cleanup_event
     mocker.patch('autosubmit.platforms.platform.Platform.get_mp_context', return_value=mp.get_context('fork'))
     local.keep_alive_timeout = 2
     max_items = 1
-    ctx = local.get_mp_context()
-    local.prepare_process(ctx)
-    local.recovery_queue = CopyQueue(ctx=ctx)
+    local.prepare_process()
+    local.recovery_queue = CopyQueue(ctx=local.ctx)
     local.cleanup_event.set() if cleanup_event else local.cleanup_event.clear()
     local.work_event.set() if work_event else local.work_event.clear()
     if recovery_queue_full:
@@ -216,9 +213,8 @@ def test_wait_mandatory_time(prepare_test, local, as_conf, mocker, cleanup_event
 def test_unique_elements(local, mocker):
     mocker.patch('autosubmit.platforms.platform.Platform.get_mp_context', return_value=mp.get_context('fork'))
     max_items = 3
-    ctx = local.get_mp_context()
-    local.prepare_process(ctx)
-    local.recovery_queue = CopyQueue(ctx=ctx)
+    local.prepare_process()
+    local.recovery_queue = CopyQueue(ctx=local.ctx)
     for i in range(max_items):
         local.recovery_queue.put(Job(f'rng{i}', f'000{i}', Status.COMPLETED, 0))
     for i in range(max_items):
@@ -313,3 +309,20 @@ def test_create_a_new_copy(local, pjm, slurm, ps_platform, ecaccess):
     assert slurm.create_a_new_copy().name == slurm.name
     assert ps_platform.create_a_new_copy().name == ps_platform.name
     assert ecaccess.create_a_new_copy().name == ecaccess.name
+
+
+def test_refresh_log_recovery_process(local, autosubmit, as_conf, mocker):
+    as_conf.misc_data["AS_COMMAND"] = 'run'
+
+    assert local.work_event is None
+
+    local.clean_log_recovery_process()
+
+    assert local.work_event is not None
+
+    p = local.work_event
+
+    with mocker.patch('multiprocessing.process.BaseProcess.is_alive', return_value=True):
+        autosubmit.refresh_log_recovery_process(platforms=[local], as_conf=as_conf)
+        assert p == local.work_event
+        assert local.work_event.is_set()

--- a/test/unit/platforms/test_locplatform.py
+++ b/test/unit/platforms/test_locplatform.py
@@ -16,9 +16,9 @@
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
 """Unit tests for the Local Platform."""
+from pathlib import Path
 
 import pytest
-from pathlib import Path
 
 from autosubmit.job.job import Job
 from autosubmit.job.job_common import Status

